src/utils/archive.rs

```diff
@@ -0,0 +1,1828 @@
+                    prev,
+                }))
+            }
+            prev = Some(pos);
+            start = header.next;
+        }
+        Ok(None)
+    }
+
+    /// Finds empty space large enough to contain the given data.
+    ///
+    /// Returns `None` if no such space can be found. Otherwise returns
+    /// the object header of the empty space and the starting position.
+    fn find_empty(
+        &self, name: &[u8], data: &[u8]
+    ) -> Result<Option<(ObjectHeader, NonZeroU64)>, ArchiveError> {
+        let mut start = self.get_empty_index()?;
+        if start.is_none() {
+            return Ok(None)
+        }
+        let size = Self::object_size(name, data);
+        let mut candidates = Vec::new();
+        while let Some(pos) = start {
+            let header = ObjectHeader::read(&self.file, pos.into())?;
+            start = header.next;
+            if Self::fits(header.size, size) {
+                candidates.push((header, pos));
+            }
+        }
+        if candidates.is_empty() {
+            return Ok(None)
+        }
+        candidates.sort_by(|left, right| left.0.size.cmp(&right.0.size));
+        Ok(candidates.first().copied())
+    }
+
+    /// Writes an object.
+    fn write_object(
+        &mut self, start: u64,
+        head: ObjectHeader, name: &[u8], meta: &Meta, data: &[u8]
+    ) -> Result<u64, ArchiveError> {
+        self.file.write(start, |write| {
+            head.write_into(write)?;
+            write.write(name)?;
+            meta.write(write)?;
+            write.write(data)?;
+            Ok(write.pos()?)
+        })
+    }
+
+    /// Returns the size of an object with the given name and content.
+    fn object_size(name: &[u8], data: &[u8]) -> u64 {
+          ObjectHeader::SIZE
+        + usize_to_u64(name.len())
+        + usize_to_u64(Meta::SIZE)
+        + usize_to_u64(data.len())
+    }
+
+    /// Returns whether an object fits into a given space.
+    ///
+    /// Specifically, checks that an object of a total size of `object_size`
+    /// (i.e., including header and name and meta) fits into empty space of
+    /// a total size of `empty_size`. This is true if they are the same or
+    /// if there is enough space left to add an empty object.
+    fn fits(empty_size: u64, object_size: u64) -> bool {
+        // Either the object fits exactly or there is enough space to add
+        // an object header
+           empty_size == object_size
+        || empty_size >= object_size + ObjectHeader::SIZE
+    }
+}
+
+
+/// # Access to the Index
+///
+impl<Meta> Archive<Meta> {
+    /// The size of a single bucket.
+    ///
+    /// This is equal to the size of the integer type we are using for archive
+    /// positions, i.e., `u64`.
+    const BUCKET_SIZE: usize = mem::size_of::<u64>();
+
+    /// Returns the hash value for a given name.
+    ///
+    /// The returned value will already be taken module the number of buckets.
+    fn hash_name(&self, name: &[u8]) -> u64 {
```

Comment:
Yes, modulo. I extended the comment to say that this is really the bucket index, not just the hash.

(I don’t want to rename the method, though, because I can’t think of a good name …)